rim/static/rim/js/checkout.js

```diff
@@ -0,0 +1,507 @@
+            if(values[current_val] > 1) {
+                $(this).toggleClass("duplicate", true);
+            } else {
+                $(this).toggleClass("duplicate", false);
+            }
+        })
+    }
+
+    function restoreData(){
+        
+        var row_count = 0;
+        // Determine how many rows are needed.
+        do{
+            var id = 'checkout_'+ $('#client').attr('id') + '_'+row_count;
+            var text = localStorage.getItem(id);
+            var next_row = $('.client textarea[type=text]').eq(row_count+1);
+
+            // If data exists and row does not exist, insert new row.
+            if ((text != null && next_row.length == 0)){
+                insert_row();
+            }
+            row_count  ++;
+        }
+        while(text != null);
+
+        // Fill the rows with data from localStorage. 
+        $('.copyable').each(function(){
+            row_index = 0;
+
+            var text_area = $(this).find('textarea[type=text]').eq(row_index);
+            var text_area_all = $(this).find('textarea[type=text]');
+
+            $(text_area_all).each(function(){
+                row_index = $(this).closest('.copyable').find('textarea[type=text]').index(this);
+                var current_class = $(text_area).closest('.copyable').attr('id');
+                var id = 'checkout_'+ current_class+ '_'+row_index;
+                var text = localStorage.getItem(id);
+                $(this).val(text);
+
+                // If data is in the client column, verify the M number.
+                if (current_class == "client"){
+                    var client_textarea = {target:this};
+                    verifyMnumber(client_textarea);
+                }
+            })
+
+            // Detect duplicates in restored data
+            var col_textarea = {target:text_area};
+            findDuplicates(col_textarea);
+
+        })     
+    }
+
+    // Function which updates the local storage for all textareas.
+    function updateStorage(){
+
+        // Clear the old values for new values
+        localStorage.clear();
+        var last_element = $('.copyable textarea[type=text]').last();
+        var last_element_index = $(last_element).closest('.copyable').find('textarea[type=text]').index(last_element);
+        var row_index = 0;
+        var save_index = 0;
+
+        // While not the last line
+        while (row_index <= last_element_index){
+            
+            empty_flag = true;      // Set a flag for each row. Assume row is empty.
+
+              $('.copyable').each(function(){
+                    
+                    var text_area = $(this).find('textarea[type=text]').eq(row_index);
+                    var current_class = $(text_area).closest('.copyable').attr('id');
+                    var id = 'checkout_'+ current_class + '_'+save_index;
+                    var last_textarea = $('#room').find('textarea[type=text]').eq(row_index);
```

Comment:
I would prefer that you use something like `$('.copyable:last textarea[type=text]').eq(row_index)` to get the last column's textarea instead of hardcoding it to look in `#room` - just in case we ever move things around.